report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java

```diff
@@ -0,0 +1,236 @@
+	void addLabel(final Label label) {
+		currentLabel.add(label);
+		if (!LabelInfo.isSuccessor(label)) {
+			noSuccessor();
+		}
+	}
+
+	/**
+	 * Adds a new instruction. Instructions are by default linked with the
+	 * previous instruction unless specified otherwise.
+	 */
+	void addInstruction(final AbstractInsnNode node) {
+		final Instruction insn = new Instruction(currentLine);
+		final int labelCount = currentLabel.size();
+		if (labelCount > 0) {
+			for (int i = labelCount; --i >= 0; ) {
+				LabelInfo.setInstruction(currentLabel.get(i), insn);
+			}
+			currentLabel.clear();
+		}
+		if (currentInsn != null) {
+			currentInsn.addBranch(insn, 0);
+		}
+		currentInsn = insn;
+		instructions.put(node, insn);
+	}
+
+	/**
+	 * Declares that the next instruction will not be a successor of the current
+	 * instruction. This is the case with an unconditional jump or technically
+	 * when a probe was inserted before.
+	 */
+	void noSuccessor() {
+		currentInsn = null;
+	}
+
+	/**
+	 * Adds a jump from the last added instruction.
+	 *
+	 * @param target jump target
+	 * @param branch unique branch number
+	 */
+	void addJump(final Label target, final int branch) {
+		jumps.add(new Jump(currentInsn, target, branch));
+	}
+
+	/**
+	 * Adds a new probe for the last instruction.
+	 *
+	 * @param probeId index in the probe array
+	 * @param branch  unique branch number for the last instruction
+	 */
+	void addProbe(final int probeId, final int branch) {
+//		REMOVED check of probes array and instead add the probes unconditionally
+//		final boolean executed = probes != null && probes[probeId];
+		currentInsn.addBranch(true, branch);
+		coveredProbes.add(new CoveredProbe(probeId, currentInsn, branch));
+	}
+
+	/**
+	 * Returns the status for all instructions of this method. This method must
+	 * be called exactly once after the instructions have been added.
+	 *
+	 * @return map of ASM instruction nodes to corresponding {@link Instruction}
+	 * instances
+	 */
+	public void fillCache() {
+		// Wire jumps:
+		for (final Jump j : jumps) {
+			j.wire();
+		}
+
+		// ADDED
+		// Traces back all instructions that are executed before reaching a probe
+		// and stores the mapping from probe to lines in #classCoverageLookup
+		for (final CoveredProbe p : coveredProbes) {
+			Instruction instruction = p.instruction;
+			Set<Integer> coveredLines = new HashSet<>();
+			while (instruction != null) {
+				coveredLines.add(instruction.getLine());
+				instruction = getPredecessor(instruction);
+			}
+			classCoverageLookup.addProbe(p.probeId, coveredLines);
+		}
+	}
+
+	/**
+	 * ADDED
+	 * Helper to get the private field predecessor from an instruction.
+	 */
+	private Instruction getPredecessor(Instruction instruction) {
+		try {
+			Field f = instruction.getClass().getDeclaredField("predecessor");
+			f.setAccessible(true);
+			instruction = (Instruction) f.get(instruction);
+		} catch (NoSuchFieldException | IllegalAccessException e) {
+			e.printStackTrace();
```

Comment:
handle with proper logging? Also at least please log that this means we have a programming error that we have to handle here in the code (I think users can't do anything in this case, right?)

Reply:
I converted into a RuntimeError as we cannot recover from this point anyway.